### PR TITLE
[NAT-124] Add arbitrary appdata support

### PR DESCRIPTION
**Pull request by ATizik**
## Description
Adds support for storing arbitrary data

## Motivation and Context

Initially needed for DonorKey from Data Donation feature, but can be used for more use cases

## How is it being implemented?

Additional set of CRUD methods are implemented, and annotation support is added on top of tags implementation to differentiate between different types of arbitrary data. Essentially an annotation is just a tag with key "custom" and some consumer provided value, which should be used to differentiate between types. We also expose `AppDataRecord` as an analogous class to the usual `Record`, which just stores byte array of arbitrary consumer-provided data. 

## How Has This Been Tested?

Several tests were added, and parts of the implementation are already covered by tests because a lot of underlying code is reused from records. Sample also implements a simple check of upload-download appData to check that the contents and annotations match.
Manual crossplatform test with iOS-uploaded Records was done to check annotations are readable on both platforms.

Annotations are implemented as an additional parameter on new methods, and old methods are left intact for compatibility. Invoking old methods are analogous to providing empty annotation list. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have updated the changelog accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by wmontwe** [comment]
Why are there 0 tests?

File: sdk-core/src/main/java/care/data4life/sdk/model/DonorKey.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.ToJson
+import okio.Buffer
+import okio.ByteString
+import java.io.CharArrayWriter
+import java.lang.reflect.Type
+
+data class DonorKey(
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/model/DonorKey.kt:31]: This should not be part of the SDK

**Comment by ATizik** [comment]
There is one integration test in sample project, but since the final implementation wasn't settled until now I didn't write documentation/more tests

File: sdk-core/src/main/java/care/data4life/sdk/AppDataResource.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+package care.data4life.sdk
+
+import care.data4life.crypto.GCKey
+import care.data4life.sdk.model.Meta
+
+/**
+ * AppDataRecord is used to store arbitrary data, analogous to Record
+ * @see care.data4life.sdk.model.Record
+ */
+data class AppDataRecord(
```
Comment (bitPogo) [inline sdk-core/src/main/java/care/data4life/sdk/AppDataResource.kt:26]: If it is analogous to Record, why they do not share a common interface?

File: sdk-core/src/main/java/care/data4life/sdk/CryptoService.java [hunk 3]
```diff
@@ -147,6 +150,27 @@ public CryptoService(
...
+    public Single<char[]> decryptToCharArray(GCKey key, String dataBase64) {
+        return Single
+                .fromCallable(() -> base64.decode(dataBase64))
+                .flatMap(decoded -> decrypt(key, decoded))
+                .map(decrypted -> StandardCharsets.UTF_8.decode(ByteBuffer.wrap(decrypted)).array())
+                .onErrorResumeNext(error -> {
+                    error(error, "Failed to decrypt string");
+                    return Single.error((D4LException) new CryptoException.DecryptionFailed("Failed to decrypt string"));
+                });
+    }
```
Comment (bitPogo) [inline sdk-core/src/main/java/care/data4life/sdk/CryptoService.java:172]: Both could be merged with the already existing string methods, to reduce code

**Review by bitPogo — COMMENTED** [review]
Besides the obvious.

File: sdk-core/src/main/java/care/data4life/sdk/model/Record.java [hunk 4]
```diff
@@ -16,15 +16,17 @@
...
 
+import java.util.List;
 import java.util.Objects;
 
 import care.data4life.fhir.stu3.model.DomainResource;
 
 public class Record<T extends DomainResource> {
     private T fhirResource;
     private Meta meta;
+    private List<String> annotations;
```
Comment (bitPogo) [inline sdk-core/src/main/java/care/data4life/sdk/model/Record.java:27]: Is this a leftover?

**Comment by wmontwe** [comment]
this is redone in #22

**Pull request closed without merging**